zulip_bots/zulip_bots/test_lib.py

```diff
@@ -95,24 +96,52 @@ def check_expected_responses(self, expectations, expected_method='send_reply',
+                                          state_handler=current_state_handler)
 
-    def call_request(self, message, expected_method, response, state_handler):
-        # type: (Dict[str, Any], str, Dict[str, Any], Optional[StateHandler]) -> None
+    def call_request(self, message, *responses, **kwargs):
+        # type: (Dict[str, Any], *Tuple[Dict[str, Any], str], **Any) -> None
+        state_handler = kwargs.pop('state_handler', None)
         if state_handler is None:
             state_handler = StateHandler()
-        # Send message to the concerned bot
-        self.message_handler.handle_message(message, self.MockClass(), state_handler)
 
-        # Check if the bot is sending a message via `send_message` function.
-        # Where response is a dictionary here.
+        # Mock BotHandler; get instance
+        bot_handler = self.MockClass()
         instance = self.MockClass.return_value
-        if expected_method == "send_message":
-            instance.send_message.assert_called_with(response)
-        else:
-            instance.send_reply.assert_called_with(message, response['content'])
+
+        # Send message to the bot
+        try:
+            self.message_handler.handle_message(message, bot_handler, state_handler)
+        except KeyError as key_err:
+            raise Exception("Message tested likely required key {}.".format(key_err))
+
+        # Determine which messaging functions are expected
+        send_messages = [call(r[0]) for r in responses if r[1] == 'send_message']
+        send_replies = [call(message, r[0]['content']) for r in responses if r[1] == 'send_reply']
+
+        # Test that call were matching in quantity, and then in details
+        fail_template = "\nMESSAGE:\n{}\nACTUAL CALLS:\n{}\nEXPECTED:\n{}\n"
+        functions_to_test = (('send_message', instance.send_message, send_messages),
+                             ('send_reply', instance.send_reply, send_replies))
+        for version, actual, expected in functions_to_test:
+            err = None
+            try:
+                assert(len(expected) == actual.call_count)
+            except AssertionError:
+                err = ("Numbers of {} called do not match those expected ({} calls, {} expected)" +
+                       fail_template).format(version, actual.call_count, len(expected),
+                                             message, actual.call_args_list, expected)
+            else:
+                try:
+                    if len(expected) > 0:
+                        actual.assert_has_calls(expected)  # any_order = True FIXME
+                except AssertionError:
+                    err = ("Calls to {} do not match those expected" +
+                           fail_template).format(version,
+                                                 message, actual.call_args_list, expected)
+            if err is not None:
+                raise AssertionError(err)
```

Comment:
Having a second look at this - would the whole chunk here be made easier if we just called assert with the error string as an argument in both cases? We could omit the except clauses and error variable.

Reply:
Reposted this comment because the previous one was made for an outdated commit and hence collapsed.